src/webgpu/shader/execution/robust_access_vertex.spec.ts

```diff
@@ -44,7 +44,430 @@ Vertex buffer contents could be randomized to prevent the case where a previous
+      this.firstIndex,
+      this.baseVertex,
+      this.firstInstance
+    );
+  }
+
+  // Insert an indirect draw call into |pass|
+  public DrawIndirect(pass: GPURenderPassEncoder) {
+    this.BindVertexBuffers(pass);
+    pass.drawIndirect(this.GenerateIndirectBuffer(), 0);
+  }
+
+  // Insert an indexed indirect draw call into |pass|
+  public DrawIndexedIndirect(pass: GPURenderPassEncoder) {
+    this.BindVertexBuffers(pass);
+    pass.setIndexBuffer(this.indexBuffer);
+    pass.drawIndexedIndirect(this.GenerateIndexedIndirectBuffer(), 0);
+  }
+
+  // Bind all vertex buffers generated
+  private BindVertexBuffers(pass: GPURenderPassEncoder) {
+    let currSlot = 0;
+    for (let i = 0; i < this.vertexBuffers.length; i++) {
+      for (let j = 0; j < this.slotsPerBuffer; j++) {
+        pass.setVertexBuffer(currSlot++, this.vertexBuffers[i]);
+      }
+    }
+  }
+
+  // Create a vertex buffer from |vertexArray|
+  private GenerateVertexBuffer(vertexArray: Float32Array): GPUBuffer {
+    const [vertexBuffer, vertexMapping] = this.device.createBufferMapped({
+      size: vertexArray.byteLength,
+      usage: GPUBufferUsage.VERTEX,
+    });
+    new Float32Array(vertexMapping).set(vertexArray);
+    vertexBuffer.unmap();
+    return vertexBuffer;
+  }
+
+  // Create an index buffer from |indexArray|
+  private GenerateIndexBuffer(indexArray: Uint16Array): GPUBuffer {
+    const [indexBuffer, indexMapping] = this.device.createBufferMapped({
+      size: indexArray.byteLength,
+      usage: GPUBufferUsage.INDEX,
+    });
+    new Uint16Array(indexMapping).set(indexArray);
+    indexBuffer.unmap();
+    return indexBuffer;
+  }
+
+  // Create an indirect buffer containing draw call values
+  private GenerateIndirectBuffer(): GPUBuffer {
+    const indirectArray = new Int32Array([
+      this.vertexCount,
+      this.instanceCount,
+      this.firstVertex,
+      this.firstInstance,
+    ]);
+    const [indirectBuffer, indirectMapping] = this.device.createBufferMapped({
+      size: indirectArray.byteLength,
+      usage: GPUBufferUsage.INDIRECT,
+    });
+    new Int32Array(indirectMapping).set(indirectArray);
+    indirectBuffer.unmap();
+    return indirectBuffer;
+  }
+
+  // Create an indirect buffer containing indexed draw call values
+  private GenerateIndexedIndirectBuffer(): GPUBuffer {
+    const indirectArray = new Int32Array([
+      this.indexCount,
+      this.instanceCount,
+      this.firstVertex,
+      this.baseVertex,
+      this.firstInstance,
+    ]);
+    const [indirectBuffer, indirectMapping] = this.device.createBufferMapped({
+      size: indirectArray.byteLength,
+      usage: GPUBufferUsage.INDIRECT,
+    });
+    new Int32Array(indirectMapping).set(indirectArray);
+    indirectBuffer.unmap();
+    return indirectBuffer;
+  }
+}
+
+// Parameterize different sized types
+interface Type {
+  format: GPUVertexFormat;
+  size: number;
+  validationFunc: string;
+}
+
+const typeInfoMap: { [k: string]: Type } = {
```

Comment:
nit: "Type" is a little vague

```suggestion
interface VertexInfo {
  format: GPUVertexFormat;
  size: number;
  validationFunc: string;
}

const typeInfoMap: { [k: string]: VertexInfo } = {
```

Reply:
Renamed. Removing the type didn't work for me because of the line where I do `const typeInfo = typeInfoMap[p.type];`, I think I need to specify that the key is a string so I may as well write the details?